File: config/src/main/java/science/monke/ConfigLoader.java
package science.monke;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import science.monke.enitity.Config;

import java.io.IOException;
import java.io.InputStream;

public class ConfigLoader {

  private ConfigLoader() {}

  public static Config load() throws IOException {
    final String environment = System.getProperty("environment");
    return load(environment);
  }

  public static Config load(final String environment) throws IOException {
    final String configName = String.format("/%s.yaml", environment);
    final InputStream resourceAsStream = ConfigLoader.class.getResourceAsStream(configName);
    final ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
    return objectMapper.readValue(resourceAsStream, Config.class);
  }
}

File: config/src/main/kotlin/science/monke/ConfigLoader.kt
package science.monke;

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.module.kotlin.readValue
import science.monke.enitity.Config
import java.io.InputStream

class ConfigLoader {

    companion object {
        fun load(): Config {
            val region = System.getProperty("region")
            val environment = System.getProperty("environment")
            return load(region, environment)
        }

        fun load(region: String, environment: String): Config {
            val configName = "$region-$environment.yaml"
            val resourceAsStream: InputStream = this::class.java.classLoader.getResourceAsStream(configName)
                ?: throw IllegalArgumentException("Config: $configName not found")
            val objectMapper = ObjectMapper(YAMLFactory())
            return objectMapper.readValue(resourceAsStream)
        }
    }


}
